### PR TITLE
Add stage-1 emulator w/o truncation

### LINKED ISSUE #80: BC+STC Stage 1 TC Processor emulator
Implement new Stage 1 emulator following the change to BC+STC inputs:
- Order input TCs in $\phi$ in each individual module
- Pack them into $\phi$ columns / bins according to the budget allocation in each $\phi$ column and module

### PR DESCRIPTION
**Pull request by adewit**
This adds a stage 1 emulator that does not truncate the TCs, but assigns them to processing columns, at the per-module level, ordered in phi. (@jbsauvan you have mostly already seen this, but imagine you may want to comment on the details). 

Some notes:
- Does not yet use real mapping information (@EmyrClement if there is progress with accessing Andy's mapping files in CMSSW we can replace the information in `HGCalLayer1PhiOrderFwConfig.h`, depending on timescale we could leave this PR open until that is done. If we're not close to being able to integrate the true mapping information, perhaps we could change the customization for the "latest" emulator back to the truncation one, so that once it's reviewed this PR could in theory be merged transparently, without breaking anything that runs in the current version of the branch. NB the code currently expects to be able to access the TC budget per processing column, and the (channel,frame) combinations accessible by each processing column. Thinking about this now this information probably needs to be provided on a module-by-module basis, which it is not currently. This will hopefully be easy to update along with the real mapping. )
- Current packing of column/channel/frame into a single number is a bit basic (understatement), I intend to update this alongside the true mapping information (making sure we know exactly what values each of these variables can take)

- The validation done so far is entirely on simulation (checking that TCs are assigned to the appropriate col/chn/frame based on the dummy mapping information.) All this only tested with a VBF 0PU sample.

### CODE REVIEW

File: L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer1ProcessorPhiOrderFw.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  const uint32_t fpga_id = fpga_id_tcs.first;
+
+  // Configuration
+  clusteringDummy_->setGeometry(geometry());
+  const std::tuple<const HGCalTriggerGeometryBase* const, unsigned, uint32_t> configuration{
+      geometry(), sector120, fpga_id};
+  PhiOrderWrapper_->configure(configuration);
+
+  PhiOrderWrapper_->process(fpga_id_tcs.second, clusters);
+}
```
Comment (EmyrClement) [inline L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer1ProcessorPhiOrderFw.h]: Is this file needed?  It looks like it duplicates what is in [L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer1ProcessorPhiOrderFw.cc](https://github.com/hgc-tpg/cmssw/pull/110/files#diff-136033accea111ac18c9b2390e64fd760370d87d4d382814ecabbe1f07c74209)?

File: L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer1ProcessorPhiOrderFw.cc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+DEFINE_EDM_PLUGIN(HGCalBackendLayer1Factory,
+                  HGCalBackendLayer1ProcessorPhiOrderFw,
+                  "HGCalBackendLayer1ProcessorPhiOrderFw");
+
+HGCalBackendLayer1ProcessorPhiOrderFw::HGCalBackendLayer1ProcessorPhiOrderFw(const edm::ParameterSet& conf)
+    : HGCalBackendLayer1ProcessorBase(conf), conf_(conf) {
+  const edm::ParameterSet& PhiOrderParamConfig = conf.getParameterSet("phiorder_parameters");
+  const std::string& PhiOrderWrapperName = PhiOrderParamConfig.getParameter<std::string>("AlgoName");
+
+  clusteringDummy_ = std::make_unique<HGCalClusteringDummyImpl>(conf.getParameterSet("clustering_dummy_parameters"));
```
Comment (EmyrClement) [inline L1Trigger/L1THGCal/plugins/backend/HGCalBackendLayer1ProcessorPhiOrderFw.cc]: Do we need a call to `clusteringDummy_->clusterizeDummy(...)` in this processor?  If not, do we still need to create and initialize a dummy clustering algorithm?

File: L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  if (sector == 1) {
+    if (phi < M_PI and phi > 0)
+      phi = phi - (2. * M_PI / 3.);
+    else
+      phi = phi + (4. * M_PI / 3.);
+  } else if (sector == 2) {
+    phi = phi + (2. * M_PI / 3.);
+  }
+  return phi;
+}
```
Comment (EmyrClement) [inline L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc]: Is `rotatedphi` an exact copy of a function defined for other wrappers e.g. `HGCalLayer1TruncationWrapper`?  If so, we could define in one place (maybe in `HGCalTriggerTools`?) rather than redefining in multiple places.  If it's easy, we can do it in this PR, or we can open a separate PR/issue.

**Comment by EmyrClement** [comment]
@adewit thanks for this - I have made some first general comments/questions.  I suggest we don't wait for the firmware mapping info to be integrated into CMSSW before merging this PR.

**Comment by adewit** [comment]
Thanks Emyr - I may not have time to update the PR early this week, so a few brief comments here: I think you're right that there's a duplicated file; and  we don't need the dummy clusterizer in this processor (it's a hangover from the truncation processor). The rotatedphi is a copy, I'll try placing it in a single place to avoid duplication. I'll try to update the PR when I can, latest early next week.

File: L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    l1t::HGCalClusterBxCollection& clusters) const {
+  std::vector<l1t::HGCalCluster> clustersTmp;
+
+  for (std::vector<edm::Ptr<l1t::HGCalTriggerCell>>::const_iterator otc = fpga_tcs_original.begin();
+       otc != fpga_tcs_original.end();
+       ++otc) {
+    clustersTmp.emplace_back(*otc);
+  }
+
+  clusters.resize(0, clustersTmp.size());
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc]: This will cause the previously built clusters to be erased (see #108)

File: L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  for (auto& tc : fpga_tcs_out) {
+    unsigned tc_cmssw_id = tc.cmsswIndex().first;
+    if (tc_cmssw_id < fpga_tcs_original.size()) {
+      clustersTmp.at(tc_cmssw_id).setColumn(tc.column());
+      clustersTmp.at(tc_cmssw_id).setFrame(tc.frame());
+      clustersTmp.at(tc_cmssw_id).setChannel(tc.channel());
+      clustersTmp.at(tc_cmssw_id).setModule(tc.moduleId());
+      clusters.set(0, tc_cmssw_id, clustersTmp.at(tc_cmssw_id));
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc]: For the same reason as above, the cluster should be pushed back instead.

File: L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+void HGCalLayer1PhiOrderWrapper::convertAlgorithmOutputs(
+    const l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_out,
+    const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs_original,
+    l1t::HGCalClusterBxCollection& clusters) const {
+  std::vector<l1t::HGCalCluster> clustersTmp;
+
+  for (std::vector<edm::Ptr<l1t::HGCalTriggerCell>>::const_iterator otc = fpga_tcs_original.begin();
+       otc != fpga_tcs_original.end();
+       ++otc) {
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc:73]: It would be more compact to use a range-based for loop:
```suggestion
  for (const auto& otc : fpga_tcs_original) {
```

File: L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    std::vector<std::pair<l1thgcfirmware::HGCalTriggerCell, int>> tcs_per_col = assignTCToCol(theConf, sorted_tcs);
+    std::vector<std::pair<l1thgcfirmware::HGCalTriggerCell, int>> tcs_with_ccf =
+        assignTCToChnAndFrame(theConf, tcs_per_col);
+
+    for (auto& tcobj : tcs_with_ccf) {
+      tcs_out.push_back(tcobj.first);
+      int col;
+      unsigned ch;
+      unsigned fr;
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc]: Could be initialized to 0

File: L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwConfig.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            chn_frame_slots_per_col_[i].push_back(std::make_pair(i + 4, j));
+          }
+        }
+        max_tcs_per_column_.push_back(std::make_pair(i, chn_frame_slots_per_col_[i].size()));
+      }
+    }
+
+    unsigned phiSector() const { return sector120_; }
+    uint32_t fpgaID() const { return fpga_id_; }
+    unsigned getColBudgetAtIndex(unsigned theColumnIndex) const {
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwConfig.h]: The budget should in principle depend on both the column and the module

File: L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwConfig.h [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      return max_tcs_per_column_.at(theColumnIndex).second;
+    }  //Get TC budget for column at index theColumnIndex in the vector
+    int getColFromBudgetMapAtIndex(unsigned theColumnIndex) const {
+      return max_tcs_per_column_.at(theColumnIndex).first;
+    }  //Get column number at index theColumnIndex in the vector
+    unsigned getChannelAtIndex(int theColumn, unsigned theChnFrameIndex) const {
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwConfig.h]: To be checked in the mapping, but I think the channel and frame also depends on the module (in addition to the column)

File: L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        theCol, theConf.getChannelAtIndex(theCol, theChnFrameIndex), theConf.getFrameAtIndex(theCol, theChnFrameIndex));
+    theTCsWithChnFrame.push_back(std::make_pair(tc.first, thePackedCCF));
+  }
+  return theTCsWithChnFrame;
+}
+
+int HGCalLayer1PhiOrderFwImpl::packColChnFrame(int column, unsigned channel, unsigned frame)
+    const {  //temporary very dumb 3 int -> 1 int conversion, can make this much smarter
+  int packed_bin = 0;
+  packed_bin = column * 100000 + channel * 1000 + frame;
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc]: Would be better to use a more standard packing like in detids or [here](https://github.com/hgc-tpg/cmssw/blob/hgc-tpg-devel-CMSSW_14_0_0_pre1/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc#L913)

File: L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  return packed_bin;
+}
+
+void HGCalLayer1PhiOrderFwImpl::unpackColChnFrame(int packedbin, int& column, unsigned& channel, unsigned& frame) const {
+  column = packedbin / 100000;
+  channel = (packedbin - column * 100000) / 1000;
+  frame = (packedbin - column * 100000 - channel * 1000);
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/src/backend_emulator/HGCalLayer1PhiOrderFwImpl.cc]: Same comment as above

**Review by jbsauvan — COMMENTED** [review]
Thanks @adewit !
Please find below some comments. The two main comments are:
- The column budget and mapping to frames, channels should in principle depend on the module id.
- Only the output clusters in the last FPGA are sent out of the producer at the moment because the output collection is re-initialized at each round (see [here](https://validation.llrhgcal.in2p3.fr/prod/list_config.php?pr=PR110) and #108 )

**Comment by adewit** [comment]
Thanks @EmyrClement and @jbsauvan for the comments. I think I have addressed them, in particular
- fixed the issue with the clusters being resized
- the `rotatedphi` function now lives in HGCalTriggerTools
- Better packing for the col-chn-frame <-> int conversion (can be beautified) 
- Added module dependency to the budget and col-> chn/frame mapping. Currently implemented by adding an extra map (map->vector and map->(map->vector)). Probably not the most memory efficient, for the map->(map->vector) we could also pack the column and module ID into a long int, to not have a nested map. Also to note: currently I've not implemented a difference in the mapping between the column/chn/frame assignments for each module, instead I've created the objects for a single dummy module ID, and the module ID variable that is passed to the functions that extract the budget or channel and frame number is replaced with this dummy value in those functions. This is just done like this to avoid having to build the map for every possible module ID.

I'd like to run a few more tests to make sure I didn't break anything, but in the meantime any comments on the updated approach of course welcome!

**Comment by jbsauvan** [comment]
Thanks @adewit !
I will have a look at the updates.
Could you also run `code-checks` and `code-format` so that the validation histograms can be produced?
Thanks!

File: L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwImpl.h [hunk 12]
```diff
@@ -23,6 +23,13 @@ namespace l1thgcfirmware {
                  l1thgcfirmware::HGCalTriggerCellSACollection& tcs_out) const;
 
   private:
+    const int columnMask_ = 0x1F;  //
+    const int columnOffset_ = 6;   //Frame and channel both have 3 bits
+    const int channelMask_ = 7;    //3 bits for channel
+    const int channelOffset_ = 3;  //Frame has 3 bits
+    const int frameMask_ = 7;      //3 bits for frame
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwImpl.h]: 3 bits wouldn't be enough to encode the frame or the channel. 
There should be in principle 108 frames and  12 channels (when not considering duplicated links)

File: L1Trigger/L1THGCal/interface/HGCalTriggerTools.h [hunk 13]
```diff
@@ -95,6 +95,8 @@ class HGCalTriggerTools {
   };
   SubDetectorType getSubDetectorType(const DetId& id) const;
 
+  double rotatedphi(double phi, unsigned sector) const;
```
Comment (jbsauvan) [inline L1Trigger/L1THGCal/interface/HGCalTriggerTools.h]: A more explicit function name would probably be better now that it is in the general trigger tools. For instance `rotatePhiToSector0()`

**Review by jbsauvan — COMMENTED** [review]
Thanks @adewit 
I just have two remaining comments, below.

File: L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwImpl.h (hunk 12 above)
Comment (adewit) [reply L1Trigger/L1THGCal/interface/backend_emulator/HGCalLayer1PhiOrderFwImpl.h]: OK I can update these to more realistic values (I put something that matched my dummy mapping as in any case we would have to double check these values when we swap in the real mapping, but indeed better to approach more sensible values)

**Comment by EmyrClement** [comment]
Perhaps more of a comment to @jbsauvan, I see the validation was run again, which I assume is triggered after any new commits that are pushed to this PR?

And then to @adewit, can you confirm the validation plots ([here](https://validation.llrhgcal.in2p3.fr/prod/list_config.php?pr=PR110)) look as expected?  From a quick look, we now see slightly higher [number of clusters being outputted by Stage 1](https://validation.llrhgcal.in2p3.fr/prod/PR110/PR110_s1fw_s1fw/cl_n.gif) with the S1 emulation than before (`Test` and `Ref` both `s1fw` on the validation webpage), which is expected as this PR is the removing truncation (for now).

**Comment by jbsauvan** [comment]
> Perhaps more of a comment to @jbsauvan, I see the validation was run again, which I assume is triggered after any new commits that are pushed to this PR?

That's correct, the validation is rerun each time a PR is updated.

**Comment by adewit** [comment]
> 
> And then to @adewit, can you confirm the validation plots ([here](https://validation.llrhgcal.in2p3.fr/prod/list_config.php?pr=PR110)) look as expected? From a quick look, we now see slightly higher [number of clusters being outputted by Stage 1](https://validation.llrhgcal.in2p3.fr/prod/PR110/PR110_s1fw_s1fw/cl_n.gif) with the S1 emulation than before (`Test` and `Ref` both `s1fw` on the validation webpage), which is expected as this PR is the removing truncation (for now).

Yes I think this looks as I would expect; we indeed turn off the truncation so I would expect more clusters in `Test` than `Ref` when running both with `s1fw` (plus there are no differences in the physics distributions when running both with `default`, and no differences in tc and tower histograms when running both with `s1fw`)

Additionally it looks like nclusters increases most in the low-energy region (but hard to see since the ratio range is rather large), which makes sense to me since before we were truncating the low-energy tcs.

**Comment by adewit** [comment]
In the meantime I've also finished running the checks that I wanted to do on my end to convince myself it really runs as I expected; unless there are further comments I think this is then ready to go.

I'm not sure if there was consensus on merging this without the true mapping incorporated; but if we do go that route I would switch back `custom_layer1_latestfw` to the emulator with truncation.

**Comment by jbsauvan** [comment]
Thanks @adewit 
That's fine to merge on my side.

**Comment by adewit** [comment]
OK! The last commit sets latestfw back to `custom_layer1_truncationfw` - we can change it to `phiorderfw` once we have the mapping incorporated.

**Comment by jbsauvan** [comment]
Hi @EmyrClement @indra-ehep 
Are there any further comments on this PR? Is it OK to merge it?
Thanks

**Comment by indra-ehep** [comment]
Dear JB,

I have no comments.

A+

Indra
